### ISSUE TITLE
If Hub accepted timestamp in MIMI protocol is mandatory can we get rid of lastSeen?

### ISSUE THREAD

**Issue by rohanmahy** [ietf-wg-mimi/draft-ietf-mimi-content#27]

**Comment by turt2live** [comment]
isn't `lastSeen` for user delivery rather than server delivery?

**Comment by rohanmahy** [comment]
lastSeen is a sender provided indicator of message order. 

hub acceptance time is a Hub provided indicator of message order. The franking mechanism now requires the hub acceptance time be delivered to a follower client in order to work. 

If the follower providers modifies the timestamp, the message will be rejected because the franking context is incorrect. The follower cannot change the franking context to match because the franking context uses a key shared only with the members and the hub.

So we are left with four cases:
- honest sender, honest hub
- honest sender, malicious hub
- malicious sender, honest hub
- malicious sender, malicious hub

There is nothing we can do about the last case. The "interesting" cases are honest sender/malicious hub, and malicious sender/honest hub.

An honest sender can correctly set lastSeen, but if it just joined it may not have the messageId of a previous message. Once a sender has included a messageId in one of its messages, it can't appear any later in the sequence from that sender, but other sender could show the same message as seen much later in the sequence. A receiver with a maliciously early hub time and a lastSeen from a single honest sender, can't correct the problem. A receiver with a maliciously late hub time and a lastSeen from an honest sender can detect a problem.

A malicious sender can't guess a message Id to include it in lastSeen before it was seen, but it can include a message Id much later in the process. A receiver has no way to determine whether to trust the sender or the hub in this case.

**Comment by rohanmahy** [comment]
@raphaelrobert says that message ordering or dropped messages by the hub might be better signaled/prevented at the MLS layer (ex: using AppAck)

**Comment by rohanmahy** [comment]
Related to the [MIMI Threat Model Issue](https://github.com/ietf-wg-mimi/mimi-protocol/issues/93)

**Comment by tgeoghegan** [comment]
@kkohbrok to see if MLS has ideas here, or wants to address this at that layer.

**Comment by kkohbrok** [comment]
Let me try to list the individual concerns here.

I believe at a baseline, we have the agreement that we want all clients to agree on some linear sequence of (application/content) messages in a room. Hub timestamps are authoritative here: Order of message arrival at the hub dictates the sequence of messages that apps SHOULD display for the user. MIMI as a protocol can only make suggestions as to the order in which messages are displayed by the actual client implementation.

However, in the reality of our distributed system, message order is more like a directed tree that depends on which messages a sender has seen at the time a message is sent (vs. the messages that may have already reached the hub, but not the sender).

Before we discuss security implications of lastSeen, maybe we should first consider whether we actually want clients to provider more information on which messages they have seen at the time of sending. If there is a critical dependency ("Do you want ice cream?", "Do you want to kill the president?", "Yes"), then users already have "replyTo" to work with. If we add more information, that just muddies the water in terms of what clients should actually display to the user.

If we want to have the hub as central authority w.r.t. message ordering, there is not much we can do to stop it from back-dating messages. We can, however, include a mechanism that ensures that users agree on the timestamps provided by the hub. I agree with @raphaelrobert that MLS is the right layer to do that. I don't think AppAck quite does the job, though, because it doesn't speak to the odering of messages (or indeed the content).

**Comment by rohanmahy** [comment]
I want to add one more point about the Hub timestamp. When *sending* a
message, the Hub's successful response contains the timestamp. If franking
is enabled, the Hub also includes the server frank which covers that
timestamp. This substantially constrains the ways the Hub can manipulate
timestamps undetected.

thanks,
-rohan

**Comment by timokoesters** [comment]
I agree that lastSeen is too complicated. If we want prevent the hub from changing timestamps, we could allow clients to add a timestamp to the content when sending the message. Other clients can show a warning if the the hub and client timestamps differ by more than a few seconds.

**Comment by rohanmahy** [comment]
Removed after unanimous feedback. Closing
